components/tidb_query/src/codec/row.rs

```diff
@@ -0,0 +1,172 @@
+// Copyright 2019 TiKV Project Authors. Licensed under Apache-2.0.
+
+use crate::codec::{table, Datum, Result};
+use crate::expr::EvalContext;
+use tikv_util::collections::HashMap;
+use tipb::ColumnInfo;
+
+pub mod v2;
+use v2::RowDecoder;
+
+pub enum Version {
+    V1,
+    V2,
+}
+
+pub fn encode(datums: Vec<Datum>, col_ids: &[i64], version: Version) -> Result<Vec<u8>> {
+    match version {
+        Version::V1 => table::encode_row(datums, col_ids),
+        Version::V2 => v2::encode(datums, col_ids),
+    }
+}
+pub fn decode(
+    data: &mut &[u8],
+    ctx: &mut EvalContext,
+    cols: &HashMap<i64, ColumnInfo>,
+) -> Result<HashMap<i64, Datum>> {
+    let version = data[0];
+    let datums = match version {
+        0..=127 => table::decode_row(data, ctx, cols).expect("TODO: handl this"),
+        v2::CODEC_VERSION => data.decode_row(ctx, cols)?,
+        _ => unimplemented!("version not supported"),
```

Comment:
It's better to put the `flag` into panic info to make diagnose easily (if someday panic here).
```suggestion
        flag => unimplemented!("version: {} not supported", flag),
```